src/sentry/api/endpoints/organization_traces.py

```diff
@@ -84,53 +88,38 @@ def data_fn(offset: int, limit: int):
                     # The orderby is intentionally `None` here as this query is much faster
                     # if we let Clickhouse decide which order to return the results in.
                     # This also means we cannot order by any columns or paginate.
                     orderby=None,
-                    limit=per_page * serialized["maxSpansPerTrace"],
-                    limitby=("trace", serialized["maxSpansPerTrace"]),
+                    limit=per_page,
+                    limitby=("trace", 1),
                     sample_rate=sample_rate,
                     config=QueryBuilderConfig(
                         transform_alias_to_input_format=True,
                     ),
                 )
-                span_results = builder.run_query(Referrer.API_TRACE_EXPLORER_SPANS_LIST.value)
-                span_results = builder.process_results(span_results)
+                trace_results = builder.run_query(Referrer.API_TRACE_EXPLORER_SPANS_LIST.value)
+                trace_results = builder.process_results(trace_results)
 
-            fields = span_results["meta"].pop("fields", {})
-            meta = {
-                **span_results["meta"],
-                "fields": {field: fields[field] for field in serialized["field"]},
-            }
+            if not trace_results["data"]:
+                return {"data": [], "meta": {"fields": {}}}
 
-            if not span_results["data"]:
-                return {"data": [], "meta": meta}
-
-            spans_by_trace: Mapping[str, list[Mapping[str, Any]]] = defaultdict(list)
-            for row in span_results["data"]:
-                spans_by_trace[row["trace"]].append(row)
-
-            trace_spans_count = sorted(
-                [(trace, len(spans)) for trace, spans in spans_by_trace.items()],
-                key=lambda item: item[0],
-                reverse=True,
-            )[:per_page]
-
-            spans_by_trace = {trace: spans_by_trace[trace] for trace, _ in trace_spans_count}
+            trace_ids: list[str] = []
 
             min_timestamp = snuba_params.end
             max_timestamp = snuba_params.start
             assert min_timestamp is not None
             assert max_timestamp is not None
-            for spans in spans_by_trace.values():
-                for span in spans:
-                    timestamp = datetime.fromisoformat(span["timestamp"])
-                    if timestamp < min_timestamp:
-                        min_timestamp = timestamp
-                    if timestamp > max_timestamp:
-                        max_timestamp = timestamp
+
+            for row in trace_results["data"]:
+                trace_ids.append(row["trace"])
+                timestamp = datetime.fromisoformat(row["timestamp"])
+                if timestamp < min_timestamp:
+                    min_timestamp = timestamp
+                if timestamp > max_timestamp:
+                    max_timestamp = timestamp
```

Comment:
I have a feeling that one trace that's an outlier will blow up the performance of searching spans with this... 
Not for this PR, but in the future maybe we should have another query param that's `maxTraceDuration` or something and some way to ignore traces over a certain time by default

Reply:
there's no way to ignore traces by their max duration without first getting their duration which requires scanning the whole period that the trace is in but i see your point.

one possibility is to add in some additional conditions to constrain the time ranges to subintervals of the full range with something like `(timestamp > a1 AND timestamp < b1) OR (timestamp > a2 AND timestamp <b2) OR ...`

Reply:
Not yet, I think with Wills work we'll be able to get their bounds much faster (hopefully) 🤞